src/miner_val_heartbeat.erl

```diff
@@ -0,0 +1,129 @@
+-module(miner_val_heartbeat).
+
+-behaviour(gen_server).
+
+-include_lib("blockchain/include/blockchain.hrl").
+-include_lib("blockchain/include/blockchain_vars.hrl").
+
+%% API
+-export([start_link/0]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-define(SERVER, ?MODULE).
+
+-record(state,
+        {
+         address :: libp2p_crypto:address(),
+         sigfun :: libp2p_crypto:sig_fun(),
+         txn_status = ready :: ready | waiting,
+         txn_wait = 10 :: non_neg_integer(),
+         chain :: undefined | blockchain:blockchain()
+        }).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+init([]) ->
+    {ok, Address0, SigFun, _ECDHFun} = blockchain_swarm:keys(),
+    Address = libp2p_crypto:pubkey_to_bin(Address0),
+    case blockchain_worker:blockchain() of
+        undefined ->
+            erlang:send_after(500, self(), chain_check),
+            {ok, #state{address = Address,
+                        sigfun = SigFun}};
+        Chain ->
+            ok = blockchain_event:add_handler(self()),
+            {ok, #state{address = Address,
+                        sigfun = SigFun,
+                        chain = Chain}}
+    end.
+
+handle_call(_Request, _From, State) ->
+    lager:warning("unexpected call ~p from ~p", [_Request, _From]),
+    Reply = ok,
+    {reply, Reply, State}.
+
+handle_cast(_Msg, State) ->
+    lager:warning("unexpected cast ~p", [_Msg]),
+    {noreply, State}.
+
+handle_info({blockchain_event, {add_block, _Hash, _Sync, _Ledger}},
```

Comment:
What is the purpose of the waiting state? just to avoid doing extra checks for a few blocks?

Reply:
so we don't spam a new txn while waiting for the old one to land